### DISCUSSION TITLE
Interface for overriting action preferences

### DISCUSSION THREAD

**Discussion by Core447** [StreamController/StreamController#49] [General]
Hey there,

You might have noticed that while you can customize icons for actions using the asset manager, the label editor currently prevents modifications to labels set by actions, including font size and color. Similarly, the layout editor controls aspects like icon size and alignments, which you might also want to customize.

I've been trying to find ways to allow users to change default actions while retaining the ability to revert to them. The challenge arises when distinguishing between user preferences and default settings, especially when reverting changes. For instance, if an action initially used the Roboto font, you changed it to Arial, and then reverted to Roboto, the app can't determine if you're enforcing that option or restoring the action's defaults (the action might use different fonts, font sizes, font colors and so on).

To address this, StreamController needs a mechanism to differentiate between user overrides and default settings. Here are three approaches I've considered:

1. **Checkbox Indicators**: Display checkboxes for properties modified by actions.  

   **Pros**:
   - Simple to understand.
   
   **Cons**:
   - May compromise UI aesthetics.

3. **Lock Icons**: Introduce lock icons next to affected properties. Clicking the lock unlocks the property for modification and locks it again to adhere to action settings. Below is a conceptual representation (lock icon only added to the font chooser button):
   ![Lock Icon](https://github.com/StreamController/StreamController/assets/100139110/78c515e9-8c5e-4edf-aac0-00c6bb59939c)

   **Pros**:
   - Clear distinction between action and user settings

   **Cons**:
   - Potential complexity in understanding.

4. **Reset Buttons**: Incorporate reset buttons for properties altered by the user. This way, StreamController can discern whether you intend to revert to action settings.

   **Pros**:
   - Intuitive for users.
  
   **Cons**:
   - A reset button may suggest the user to just clear the setting

Each approach has its trade-offs, and we need to carefully consider usability alongside functionality. Let me know your thoughts or if you have any other suggestions.

**Let me know what you like the best and if you have any other ideas to solve this problem.**

**Comment by G4PLS** [comment]
I personally like option 2 the best. Having a Lock for every thing you can customize could look ugly though, but having it globally on a label by label action would make it so you only have "Use Action Settings" and "Only Use Custom Settings"
Generally it looks the cleanest, and when going with this approach plugins can maybe also specify if users are able to use custom settings. This way a more sensitive plugin to UI changes would also have the option to make it so the user cant mess anything up

**Reply by Core447** [reply]
Having one lock for the whole level wouldn't allow users to just change the font size or color, though

**Reply by G4PLS** [reply]
Yeah thats what i meant
Having it globally for every label would mean the user has one lock and then he has control over everything
Having it on every component the user can change (Font, Color, Text) Would make it so the user can change it very specifically but i think it could look cluttered

**Reply by Core447** [reply]
Ok, the middle ground would be a reset button instead of a lock button that only appears if you modified the property

**Reply by G4PLS** [reply]
This sounds good to me!

**Reply by v4u6h4n** [reply]
Yeah me to, they have something similar in Flatseal that looks aesthetically good.

**Reply by Core447** [reply]
The ui (for modified properties) with reset buttons will look like this:
![image](https://github.com/StreamController/StreamController/assets/100139110/979ec513-933a-4526-852d-9b210b5cd504)

**Reply by v4u6h4n** [reply]
Looks fresh to me ;-)
